src/widget/wsingletoncontainer.cpp

```diff
@@ -0,0 +1,55 @@
+#include "widget/wsingletoncontainer.h"
+
+#include <QLayout>
+
+#include "util/assert.h"
+#include "widget/wlibrary.h"
+
+// static
+WSingletonContainer::WidgetMap WSingletonContainer::m_singletons;
+
+// static
+WSingletonContainer* WSingletonContainer::getSingleton(
+        QString objectname, QWidget* pParent) {
+   WidgetMap::const_iterator widget_it = m_singletons.find(objectname);
+    if (widget_it == m_singletons.end()) {
+        qWarning() << "ERROR: Asked for an unknown singleton widget:"
+                   << objectname;
+        return NULL;
+    }
+    return new WSingletonContainer(objectname, pParent);
+}
+
+WSingletonContainer::WSingletonContainer(QString objectname, QWidget* pParent)
+        : WWidgetGroup(pParent), m_pWidget(NULL) {
+    WidgetMap::const_iterator widget_it = m_singletons.find(objectname);
+    DEBUG_ASSERT(widget_it != m_singletons.end());
```

Comment:
Hm, why not just pass the widget directly in the constructor? 

It's fine to keep the assert but future developers may not be aware of that invariant. It's best to design this class so that it doesn't segfault in unexpected conditions (i.e. check for null m_pWidget before dereferencing, don't dereference an end() iterator, etc.).